Move to libglvnd for OpenGL

abbradar opened the pull request:
###### Motivation for this change

* ~~Removes somewhat need for LD_LIBRARY_PATH in NixOS (it's still left for other usecases but it would be much more simple to get rid of it now)~~ (actually not because GLX vendor implementations are searched by name -- it would be needed to patch those too. Still, at least no libraries should be overridden now with `LD_LIBRARY_PATH` in usual case, it should be only used for `dlopen` search);
* Shrinks closure size of both `libGL` and `mesa_noglu.out`;
* ~~Fixes https://github.com/NixOS/nixpkgs/issues/9415 with newer video drivers.~~ (actually not because `libc`/`libstdc++` would still mix -- it needs something akin to `libcapsule`. Would help with simple cases though)

Besides doing the move itself I've done two more changes:

* Dropped X libraries from our `display-manager.service`'s `LD_LIBRARY_PATH`; see commit message for why was it done. tl;dr: it seems to not be needed anymore so this is a cleanup.
* Added `virgl` driver to Mesa and OpenGL support to Qemu. This still doesn't work (Qemu segfaults when using OpenGL support and trying to get virgl acceleration) but hopefully we won't need another mass rebuild to fix this. Initially I wanted to use it for testing this change.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

Tested on:

1. A system with unfree NVIDIA drivers (my desktop, full configuration);
2. A system with Mesa drivers (my laptop, full configuration);
3. Steam;
4. Bumblebee.

This can break packages; usual fix is to add `mesa_noglu` (for `gbm`, `dri` and other dependencies) or `libdrm` (it was propagated before I think).

cc @vcunat (I remember you tried this some time ago).

abbradar committed:
libglvnd: 2016-12-22 -> 1.0.0

Move driverLink attribute there.

abbradar committed:
mesa_noglu: build with libglvnd

abbradar committed:
mesa_noglu: build virgl driver

It's used for 3D acceleration inside qemu.

abbradar committed:
treewide: refactor to use libglvnd

* Implement libGL as a symlink package which uses libraries from libglvnd and
  headers from Mesa (since ones from libglvnd are outdated).
* Use libGL_driver.driverLink treewide; add FHS paths where possible.

abbradar committed:
qt4: remove OpenGL absolute path patch

In resulting derivation there's no mention of this path anyway (checked with grep).

abbradar committed:
nvidia service: rely on libglvnd

abbradar committed:
opengl service: move options under common hardware.opengl tree

Purely cosmetic.

abbradar committed:
virglrenderer: add libdrm as dependency

abbradar committed:
qemu: add virgl renderer support

abbradar committed:
glxinfo: fix build

abbradar committed:
xorg: update autogeneration script

* Fix shebang so that it's usable on NixOS;
* Enable warnings (they were enabled with Perl flag before);
* Switch from mesa to libGL.

abbradar committed:
qemu: use SDL2 as SDL library

This is needed for OpenGL support.

abbradar committed:
qemu: add OpenGL support

abbradar committed:
xserver service: cleanup LD_LIBRARY_PATH

X libraries in LD_LIBRARY_PATH seem to not be needed anymore.
I've tracked this addition as far as I could
(02cef04c8187f12fd160d5fedf0ec8a928f02639) and they seem to be added for unfree
NVIDIA and ATI drivers but at least for NVIDIA they are not needed anymore. We
can add them with patchelf instead if it turns out to be the case with ATI.

abbradar committed:
xorg.xorgserver: don't use XORG_DRI_DRIVER_PATH

Our patch which adds this envvar has been removed.

abbradar committed:
qemu-vm service: quote arguments in qemu runner

abbradar committed:
gtk3: fix build

abbradar committed:
qt5.qtwebengine: fix build

abbradar committed:
fbida: fix build

abbradar committed:
cogl: fix build

abbradar committed:
intel-gpu-tools: enable parallel building

abbradar committed:
opengl service: don't override mesa non-drivers in runtime

abbradar committed:
primusLib: fix path to libglapi

abbradar committed:
steam: add libGL to the environment

Also remove duplicate libGLU.

abbradar committed:
bumblebee: remove X libraries from LD_LIBRARY_PATH

They seem to no longer be needed.

abbradar commented:
To boast a little:

```
➜  ~ ls /run/opengl-driver/lib | grep libGL
libGLX_mesa.la
libGLX_mesa.so
libGLX_mesa.so.0
libGLX_mesa.so.0.0.0
➜  ~ glxinfo | grep version                
server glx version string: 1.4
client glx version string: 1.4
GLX version: 1.4
    Max core profile version: 4.5
    Max compat profile version: 3.0
    Max GLES1 profile version: 1.1
    Max GLES[23] profile version: 3.1
OpenGL core profile version string: 4.5 (Core Profile) Mesa 17.3.6
OpenGL core profile shading language version string: 4.50
OpenGL version string: 3.0 Mesa 17.3.6
OpenGL shading language version string: 1.30
OpenGL ES profile version string: OpenGL ES 3.1 Mesa 17.3.6
OpenGL ES profile shading language version string: OpenGL ES GLSL ES 3.10
➜  ~ 
```

xeji commented:
Qemu virtio graphics: see also #35564

abbradar commented:
Ahaha, it seems I've even seen your PR but then forgot about it and went ahead reimplementing. I'll compare them when I'm near a laptop again.

xeji commented:
Essentially the same. I'll close and you get to fix the bugs :-)

abbradar commented:
@oxij After some thought this doesn't always help with non-NixOS even for "modern drivers" case because host and nixpkgs `libc`/`libstdc++` can be of different versions (which I completely forgot). Nothing short of `libcapsule`/`dlmopen` would help.

ttuegel reviewed (approved):
The Qt 5 parts look fine to me.

dezgeg commented on nixos/modules/services/x11/xserver.nix:
```diff
@@ -640,9 +640,7 @@ in
-            XORG_DRI_DRIVER_PATH = "/run/opengl-driver/lib/dri"; # !!! Depends on the driver selected at runtime.
-            LD_LIBRARY_PATH = concatStringsSep ":" (
-              [ "${xorg.libX11.out}/lib" "${xorg.libXext.out}/lib" "/run/opengl-driver/lib" ]
+            LD_LIBRARY_PATH = concatStringsSep ":" ([ "/run/opengl-driver/lib" ]
```
Redundant code here.

oxij commented:
> because host and nixpkgs `libc`/`libstdc++` can be of different versions (which I completely forgot)

Ah, you're right. That's unfortunate. libcapsule seems to be half-finished and abandoned and I'm not particularly exited about the idea of finishing the work myself. It's not my itch. And I'd prefer non-NixOS users to simply use drivers built by nix anyway. I.e. the only non-pure thing nix-built binaries would use on non-NixOS would be the kernel.

Also, it's not like this PR solves the need for `LD_LIBRARY_PATH`/`LD_FALLBACK_PATH` on non-NixOS anyway because `driverLink` is NixOS-specific.

Hence, this PR and those other issues are separate issues. Hence, I'll proceed as planned before with my itches.

As to this PR, everything except ad1d918dc709a2e5f6ef51b2e8bbec7abc311654 LGTM (but I have not run it yet). We ship `nvidiaLegacy*` and old AMD drivers in nixpkgs (and I use them on some of my machines), hence those should be fixed before merging the patch, I think.

Also, `libGL`/`mesa_noglu` usage got really confusing with the `.stubs` thing, we just had the `mesa -> libGL` tree-wide rewrite, and now this PR does the reverse for some packages. (I think I understand why, but it got confusing for a second.) I think this needs some clear documentation somewhere near that `libGL` attr in `pkgs/top-level/all-packages.nix` that explains which attr to use when.

dezgeg commented:
> Also, it's not like this PR solves the need for `LD_LIBRARY_PATH`/`LD_FALLBACK_PATH` on non-NixOS anyway because `driverLink` is NixOS-specific.

Why not? libglvnd is configured to look into non-NixOS paths here: https://github.com/NixOS/nixpkgs/pull/37369/files#diff-9179be247d2ec19fbbf4b85e1ac99692R22

oxij commented:
> Why not? libglvnd is configured to look into non-NixOS paths here: https://github.com/NixOS/nixpkgs/pull/37369/files#diff-9179be247d2ec19fbbf4b85e1ac99692R22 

That's only for `libva`, is it not? Seems like a hack that follows from the lack of `LD_FALLBACK_PATH` to me.

oxij commented:
Ah, hm. You probably meant a different chunk from what GitHub has shown me without JS.

> "-DDEFAULT_EGL_VENDOR_CONFIG_DIRS=\"${driverLink}/share/glvnd/egl_vendor.d:/etc/glvnd/egl_vendor.d:/usr/share/glvnd/egl_vendor.d\""

is ok, but

- that assumes a distro that already does `libglvnd`,
- doesn't solve the libcapsule problem.

To be honest, I would remove all FHS/non-NixOS hacks from this PR so that people would not start to depend on this not-entirely-correct-not-always-working behavior and do everything `nixGL`-style.

abbradar commented:
@ttuegel @dezgeg @oxij  Thanks for the review!

@oxij

> Ah, you're right. That's unfortunate. libcapsule seems to be half-finished and abandoned and I'm not particularly exited about the idea of finishing the work myself. It's not my itch.

Nah, I didn't even considered you would implement it ^^" Just mentioned this to you since it was discussed before.

> And I'd prefer non-NixOS users to simply use drivers built by nix anyway. I.e. the only non-pure thing nix-built binaries would use on non-NixOS would be the kernel.

As @dezgeg said, unfortunately this is a problem for non-free drivers.

> As to this PR, everything except ad1d918 LGTM (but I have not run it yet). We ship `nvidiaLegacy*` and old AMD drivers in nixpkgs (and I use them on some of my machines), hence those should be fixed before merging the patch, I think.

`nvidia` (except really old ones) should be built by the generic builder nowadays so if it works with the latest driver it should work at least up to these. But as you noticed I haven't tried this on except latest nvidia drivers so yeah, this might need further fixing for AMD/very old nvidia. It's too hacky as is though, better to fix this in the drivers.

> - that assumes a distro that already does `libglvnd`, - doesn't solve the libcapsule problem. To be honest, I would remove all FHS/non-NixOS hacks from this PR so that people would not start to depend on this not-entirely-correct-not-always-working behavior and do everything `nixGL`-style.

Hm, I usually prefer things in order "don't work at all" < "work sometimes but you'll need a documented fix for others" < "don't work at all with clear instructions _when attempted_ what do you need to do" < "just work" so it seems better to me than current state (I don't think we can make binaries output a warning prompting for `nixGL`).

dezgeg commented:
Well, the "libcapsule problem" isn't solveable in any way, except by using libcapsule. The problem of a Nixpkgs app on non-NixOS loading a non-Nixpkgs OpenGL driver are the exactly the same problems that happen when a Nixpkgs x.y app loading a NixOS (x+1).z OpenGL driver. Or a proprietary game depending on a vendored dynamically linked libstdc++ on a NixOS OpenGL driver, for that matter.

oxij commented:
@abbradar

> As @dezgeg said, unfortunately this is a problem for non-free drivers.

Why? `nixGL`-style wrapper can wrap any version of the driver, including exactly the same version as used by the distro.

> It's too hacky as is though, better to fix this in the drivers.

I agree, but making someone's machine to boot into tty instead of X would be unfortunate. I'll check all the drivers I can with this branch later this week.

> I don't think we can make binaries output a warning prompting for `nixGL`

We can't. But we can make `nix-env` to print a warning when installing and updating packages and we can make a version of `buildEnv` that wraps all binaries.

A simple idea how that can be done without patching `nix`: just `trace` warnings in stdenv's `mkDerivation` when evaluating without `config.binariesWillRunInNixOSLikeEnvironment` (or `config.IKnowIShouldUseTheWrapper`, or whatever) option (and set that option in NixOS by default). (And I wanted to implement such warnings for https://github.com/NixOS/nixpkgs/issues/36229#issuecomment-370241125 anyway).

> order of preference

If things never work without you doing the proper thing (the wrapper) then (if you didn't already see the warning described above) you would simply look up the "no libGL" error message and find #9415/the documentation. Meanwhile, with the FHS hacks in place errors are going to be wildly different depending on the user's config/installed libraries.

@dezgeg

> ... Nixpkgs x.y app loading a NixOS (x+1).z OpenGL driver. Or a proprietary game depending on a vendored dynamically linked libstdc++ on a NixOS OpenGL driver, for that matter

Hm, right. I guess, strictly speaking, anything installed outside of `systemPackages` can have the same problem on a given booted system. Including `nix-env -i`-installed packages from a different revision of nixpkgs.

But, if `libstdc++`s (or any other library for that matter) disagrees with another version of itself we can't really do anything without libcapsule. And even with libcapsule stuff can break (think two versions libstdc++ doing their own buffering on the same file descriptor), just with a smaller probability.

@abbradar, @descriptor

Personally, I would be ok with leaving the hacks in (I don't run nix things on non-NixOS, nor do I run any unfree software except for the drivers), but it still seems wise to me to push as much of responsibility for consistency into nixpkgs and make all inconsistent use cases to fail with a deterministic error message when possible.

I would even go as far as to replace the `driverLink` thing with a env variable NixOS would set by default. This way NixOS would use the same code path as non-NixOS with `nixGL`-style wrapper and non-NixOS problems like #9415 would simply be impossible. This also would alleviate

> Nixpkgs x.y app loading a NixOS (x+1).z OpenGL driver

somewhat by allowing you to use different `nixGL`-style wrappers for NixOS itself.

oxij commented:
Btw, would be nice if someone here could look at the somewhat related #35115 of mine. It got ignored for a while now.

abbradar commented:
@oxij 

> Why? `nixGL`-style wrapper can wrap any version of the driver, including exactly the same version as used by the distro.

That's not always possible -- host distribution can run any given version of a driver, not just from our given set. Even if they use some "latest" driver "latest" is in desync with our "latest" so it would frequently break on updates of either nixpkgs or host OS.

> I agree, but making someone's machine to boot into tty instead of X would be unfortunate. I'll check all the drivers I can with this branch later this week.

Much appreciated! I only have Intel and modern Nvidia machines. You'd want to check `amdgpu-pro`, `ati_unfree` (not sure it works in NixOS) and a very old `legacy_173` nvidia driver -- the rest of NVIDIA should be covered by the generic builder (not immune of course).

> We can't. But we can make `nix-env` to print a warning when installing and updating packages and we can make a version of `buildEnv` that wraps all binaries. A simple idea how that can be done without patching `nix`: just `trace` warnings in stdenv's `mkDerivation` when evaluating without `config.binariesWillRunInNixOSLikeEnvironment` (or `config.IKnowIShouldUseTheWrapper`, or whatever) option (and set that option in NixOS by default). (And I wanted to implement such warnings for #36229 (comment) anyway).

That's painful for the vast majority of users who _don't_ use OpenGL, but the idea is good -- we can show warnings if something in `meta` exists, like `glUsed = true;` (and this `config` value is not set). I like this but not yet sure removing standard paths from our libraries is a good idea for reasons of proprietary drivers above.

> If things never work without you doing the proper thing (the wrapper) then (if you didn't already see the warning described above) you would simply look up the "no libGL" error message and find #9415/the documentation. Meanwhile, with the FHS hacks in place errors are going to be wildly different depending on the user's config/installed libraries.

Agreed but combined with warnings it could be better than just warnings -- people still know it can be problematic but sometimes it just works. Actually I won't expect much problems with proprietary drivers because they are usually minimal in their use of `libc` and don't use `libstdc++` at all. Most problems came from Radeon open-source drivers because they use `libstdc++` and LLVM extensively (not that it's a bad thing!). Also, as @dezgeg noticed, we already do this with Nix-on-NixOS too with OpenGL drivers and also with e.g. Qt modules.

> I would even go as far as to replace the `driverLink` thing with a env variable NixOS would set by default. This way NixOS would use the same code path as non-NixOS with `nixGL`-style wrapper and non-NixOS problems like #9415 would simply be impossible.

Good idea actually --right now it'd be much more difficult to set all necessary paths for `nixGL`. Maybe I'll look into this (not in the near future though).

oxij commented:
> host distribution can run any given version of a driver

True, but it's fairly easy to have several checkouts of nixpkgs tree.

> I only have Intel and modern Nvidia machines. You'd want to check `amdgpu-pro`, `ati_unfree` (not sure it works in NixOS) and a very old `legacy_173` nvidia driver -- the rest of NVIDIA should be covered by the generic builder (not immune of course).

I can check `ati_unfree` (it works if you patch NixOS `xserver` module a bit) and, possibly, `nvidiaLegacy173`. I don't have anything that runs `amdgpu-pro`.

> we can show warnings if something in `meta` exists, like `glUsed = true;` (and this `config` value is not set)

Nice. Actually, we can add that warning to `libGL` itself so that it would print it only when you install an app that uses it.

> Agreed but combined with warnings it could be better than just warnings -- people still know it can be problematic but sometimes it just works.

Okay, I see your point.

> Good idea actually --right now it'd be much more difficult to set all necessary paths for `nixGL`. Maybe I'll look into this (not in the near future though).

Do you mean this patchset will also break `nixGL` (not just -style, the original thing)? That would be unfortunate too.

abbradar commented:
About `nixGL` - it shouldn't be affected since NVIDIA unfree drivers already worked there and they use libglvnd. 
-- 
Nikolay.

abbradar commented:
Testing and review much appreciated! I've rebased pushed our combined PRs to staging.

abbradar commented:
@oxij To answer an old comment which I forgot to address earlier -- I didn't document `mesa_noglu` vs `libGL` because there's not much change in usage actually -- `mesa_noglu` is only needed when there're specific Mesa libraries you want to use (`libwayland-egl` etc.).

I've also moved `libva-full` to `aliases.nix` and updated `libva1` to use the same `{,-minimal}` convention.

gnidorah commented:
cc @avnik

gnidorah commented on pkgs/development/libraries/mesa/default.nix:
```diff
@@ -185,10 +188,22 @@ stdenv.mkDerivation {
-    # move share/vulkan/icd.d/
+    substituteInPlace "$dev/lib/pkgconfig/dri.pc" --replace '$(drivers)' "${libglvnd.driverLink}"
+
+    # remove GLES libraries; they are provided by libglvnd
```
You may do this instead https://github.com/NixOS/nixpkgs/pull/34282/files#diff-481f6f0a7a36656c0dc31b9ccd2202afL100

abbradar replied:
Would that still build GLES support itself?

avnik replied:
Anyway, libglvnd's GLES is dispatcher only, it need real GLES somewhere

abbradar replied:
I was baffled by this too but `glxinfo` [shows working OpenGL ES](https://github.com/NixOS/nixpkgs/pull/37369#issuecomment-374176822). My guess is that `--enable-gles*` builds actual GLES into `lib{GLX,EGL}_mesa` and Mesa's `libGLES` are wrappers somewhat similar to libglvnd's. This isn't backed by anything though. I used Arch Linux package for ~stealing ideas~inspiration and they also use `--enable-gles` and manually install only needed libraries later.

oxij commented:
> Testing and review much appreciated! I've rebased pushed our combined PRs to staging.
> I've also moved `libva-full` to `aliases.nix` and updated `libva1` to use the same `{,-minimal}` convention.

Awesome! Many thanks.

> @oxij To answer an old comment which I forgot to address earlier -- I didn't document `mesa_noglu` vs `libGL` because there's not much change in usage actually -- `mesa_noglu` is only needed when there're specific Mesa libraries you want to use (`libwayland-egl` etc.).

I'm not sure that's entirely correct, because you can, in principle, try to `-lGL` with `libGL` stubs and fail while not requiring any particular `libGL`.

abbradar commented:
@oxij Not sure I understood you, do you mean it's possible `libglvnd` doesn't export all GL symbols for some latest standard? I don't think that'll happen in practice -- `libglvnd` is used by NVIDIA's driver to export symbols and it's largely autogenerated.

oxij commented:
Ah, I see. It indeed is autogenerated from the XML OpenGL spec files (https://github.com/NVIDIA/libglvnd/blob/master/src/generate/gen_libOpenGL_exports.py). Then that's not an issue. Thanks for clarification.

Mic92 commented:
Is this indented that I have to add mesa_noglu to xorgserver now? 
Currently it is broken: https://hydra.nixos.org/build/74046691

andrewrk commented:
This broke glfw and/or libepoxy, whichever one tries to find libGL.so.1. It also broke zsnes.

avnik commented:
On Sat, Sep 08, 2018 at 07:48:07PM -0700, Andrew Kelley wrote:
> This broke glfw and/or libepoxy, whichever one tries to find libGL.so.1. It also broke zsnes.
> 

Looks like following 5 lines need to be patched with full paths to
libglvnd libraries -- https://github.com/anholt/libepoxy/blob/master/src/dispatch_common.c#L192
(all other indirection/dispatching would be done there)

andrewrk commented:
If there's a static executable that wants to work on all Linux platforms, how can it find OpenGL in such a way that it will find the correct version on NixOS?

Mic92 commented:
Static executables do not load dynamic libraries. Dynamically linked libraries either load libraries from LD_LIBRARY_PATH environment variable or the rpath that is set in the executable (the latter one can be adjusted with [patchelf](https://nixos.org/patchelf.html))

andrewrk commented:
static executables can load dynamic libraries just fine. Here's a test case covering this behavior: https://github.com/ziglang/zig/tree/7c9f7b72c59e7c6de38038f512ae332fc164e8d7/test/standalone/load_dynamic_library

on my system,
`LD_LIBRARY_PATH=/run/opengl-driver/lib:/run/opengl-driver-32/lib`

in this dir:

```
bellagio              libtxc_dxtn.so         libXvMCgallium.so.1
d3d                   libtxc_dxtn.so.0       libXvMCgallium.so.1.0.0
dri                   libtxc_dxtn.so.0.0.0   libXvMCnouveau.so
libEGL_mesa.la        libvulkan_intel.la     libXvMCnouveau.so.1
libEGL_mesa.so        libvulkan_intel.so     libXvMCnouveau.so.1.0
libEGL_mesa.so.0      libvulkan_radeon.la    libXvMCnouveau.so.1.0.0
libEGL_mesa.so.0.0.0  libvulkan_radeon.so    libXvMCr600.so
libGLX_mesa.la        libxatracker.la        libXvMCr600.so.1
libGLX_mesa.so        libxatracker.so        libXvMCr600.so.1.0
libGLX_mesa.so.0      libxatracker.so.2      libXvMCr600.so.1.0.0
libGLX_mesa.so.0.0.0  libxatracker.so.2.3.0  pkgconfig
libtxc_dxtn.la        libXvMCgallium.so      vdpau
```

how is code supposed to know which one to load?

oxij commented:
> This broke glfw and/or libepoxy, whichever one tries to find libGL.so.1. It also broke zsnes.

Can you give examples in "I run this, I get this error" format so I could try to reproduce those?

> https://github.com/anholt/libepoxy/blob/master/src/dispatch_common.c#L192

Can they just run pkg-config or something? Hardcoding library names with their versions looks rather strange to me.

avnik commented:
On Sun, Sep 09, 2018 at 04:02:54PM -0700, Jan Malakhovski wrote:
> > This broke glfw and/or libepoxy, whichever one tries to find libGL.so.1. It also broke zsnes.
> 
> Can you give examples in "I run this, I get this error" format so I could try to reproduce those?
> 

I'd like to see "how-to-reproduce" as well.

> > https://github.com/anholt/libepoxy/blob/master/src/dispatch_common.c#L192
> 
> Can they just run pkg-config or something? Hardcoding library names with their versions looks rather strange to me.
> 

These names are standard and their sonames bound to api versions as
well.

Actually it have two solutions:

1. add libglvnd to `/run/opengl-drivers` (replace just link to
mesa-or-nvidia with link merged tree of all listed drivers). Personally I
prefer to have it configurable.

2. patch all GL loaders with full paths to libGL.so.1 and friends
(resolved to `${libglvnd}/lib`). Even if `libepoxy` load libGL
dynamically -- `libGL.so.1` still be explicit dependency, and full path
provide trackable source of it.

andrewrk commented:
Here's a how-to repro:  https://github.com/NixOS/nixpkgs/issues/46423#issuecomment-421789242

cx405 commented:
Just migrated to 18.09 from 18.03, and:

```
$ openarena
/nix/store/a1jmb86qk6fj3zvdmixhpwxlv1phliw6-openarena-0.8.8/openarena-0.8.8/openarena.x86_64: error while loading shared libraries: libGL.so.1: cannot open shared object file: No such file or directory

$ echo $LD_LIBRARY_PATH
/nix/store/mb471p2iqq73lfz9z41p7gbszpa3rk9r-sane-config/lib/sane:/run/opengl-driver/lib:/run/opengl-driver-32/lib

$ ls -1 /run/opengl-driver/lib
bellagio
d3d
dri
libEGL_mesa.la
libEGL_mesa.so
libEGL_mesa.so.0
libEGL_mesa.so.0.0.0
libGLX_mesa.la
libGLX_mesa.so
libGLX_mesa.so.0
libGLX_mesa.so.0.0.0
libtxc_dxtn.la
libtxc_dxtn.so
libvulkan_intel.la
libvulkan_intel.so
libvulkan_radeon.la
libvulkan_radeon.so
libxatracker.la
libxatracker.so
libxatracker.so.2
libxatracker.so.2.3.0
libXvMCgallium.so
libXvMCgallium.so.1
libXvMCgallium.so.1.0.0
libXvMCnouveau.so
libXvMCnouveau.so.1
libXvMCnouveau.so.1.0
libXvMCnouveau.so.1.0.0
libXvMCr600.so
libXvMCr600.so.1
libXvMCr600.so.1.0
libXvMCr600.so.1.0.0
vdpau
```

cx405 commented:
@Mic92 Looks like it just tries to locate the library that isn't there:
```
      2630:     file=libGL.so.1 [0];  needed by /nix/store/a1jmb86qk6fj3zvdmixhpwxlv1phliw6-openarena-0.8.8/openarena-0.8.8/openarena.x86_64 [0]
      2630:     find library=libGL.so.1 [0]; searching
      2630:      search path=/nix/store/833lmkfczmbzmxfnvhb00bl0kmb9yly3-SDL-1.2.15/lib:/nix/store/8fz98xhlk563kgsvmvihyj8m6i22pcc7-libogg-1.3.3/lib/tls/x86_64/x86_64:/nix/store/8fz98xhlk563kgsvmvihyj8m6i22pcc7-libogg-1.3.3/lib/tls/x86_64:/nix/store/8fz98xhlk563kgsvmvihyj8m6i22pcc7-libogg-1.3.3/lib/tls/x86_64:/nix/store/8fz98xhlk563kgsvmvihyj8m6i22pcc7-libogg-1.3.3/lib/tls:/nix/store/8fz98xhlk563kgsvmvihyj8m6i22pcc7-libogg-1.3.3/lib/x86_64/x86_64:/nix/store/8fz98xhlk563kgsvmvihyj8m6i22pcc7-libogg-1.3.3/lib/x86_64:/nix/store/8fz98xhlk563kgsvmvihyj8m6i22pcc7-libogg-1.3.3/lib/x86_64:/nix/store/8fz98xhlk563kgsvmvihyj8m6i22pcc7-libogg-1.3.3/lib:/nix/store/fw6kv6w6fjh5q7pmnr470pkp93ds13hb-libvorbis-1.3.6/lib/tls/x86_64/x86_64:/nix/store/fw6kv6w6fjh5q7pmnr470pkp93ds13hb-libvorbis-1.3.6/lib/tls/x86_64:/nix/store/fw6kv6w6fjh5q7pmnr470pkp93ds13hb-libvorbis-1.3.6/lib/tls/x86_64:/nix/store/fw6kv6w6fjh5q7pmnr470pkp93ds13hb-libvorbis-1.3.6/lib/tls:/nix/store/fw6kv6w6fjh5q7pmnr470pkp93ds13hb-libvorbis-1.3.6/lib/x86_64/x86_64:/nix/store/fw6kv6w6fjh5q7pmnr470pkp93ds13hb-libvorbis-1.3.6/lib/x86_64:/nix/store/fw6kv6w6fjh5q7pmnr470pkp93ds13hb-libvorbis-1.3.6/lib/x86_64:/nix/store/fw6kv6w6fjh5q7pmnr470pkp93ds13hb-libvorbis-1.3.6/lib:/nix/store/9c9vryq51hlw0jr8ddk2yk1qvkaq0vl2-curl-7.61.0/lib/tls/x86_64/x86_64:/nix/store/9c9vryq51hlw0jr8ddk2yk1qvkaq0vl2-curl-7.61.0/lib/tls/x86_64:/nix/store/9c9vryq51hlw0jr8ddk2yk1qvkaq0vl2-curl-7.61.0/lib/tls/x86_64:/nix/store/9c9vryq51hlw0jr8ddk2yk1qvkaq0vl2-curl-7.61.0/lib/tls:/nix/store/9c9vryq51hlw0jr8ddk2yk1qvkaq0vl2-curl-7.61.0/lib/x86_64/x86_64:/nix/store/9c9vryq51hlw0jr8ddk2yk1qvkaq0vl2-curl-7.61.0/lib/x86_64:/nix/store/9c9vryq51hlw0jr8ddk2yk1qvkaq0vl2-curl-7.61.0/lib/x86_64:/nix/store/9c9vryq51hlw0jr8ddk2yk1qvkaq0vl2-curl-7.61.0/lib:/nix/store/mb471p2iqq73lfz9z41p7gbszpa3rk9r-sane-config/lib/sane/tls/x86_64/x86_64:/nix/store/mb471p2iqq73lfz9z41p7gbszpa3rk9r-sane-config/lib/sane/tls/x86_64:/nix/store/mb471p2iqq73lfz9z41p7gbszpa3rk9r-sane-config/lib/sane/tls/x86_64:/nix/store/mb471p2iqq73lfz9z41p7gbszpa3rk9r-sane-config/lib/sane/tls:/nix/store/mb471p2iqq73lfz9z41p7gbszpa3rk9r-sane-config/lib/sane/x86_64/x86_64:/nix/store/mb471p2iqq73lfz9z41p7gbszpa3rk9r-sane-config/lib/sane/x86_64:/nix/store/mb471p2iqq73lfz9z41p7gbszpa3rk9r-sane-config/lib/sane/x86_64:/nix/store/mb471p2iqq73lfz9z41p7gbszpa3rk9r-sane-config/lib/sane:/run/opengl-driver/lib/tls/x86_64/x86_64:/run/opengl-driver/lib/tls/x86_64:/run/opengl-driver/lib/tls/x86_64:/run/opengl-driver/lib/tls:/run/opengl-driver/lib/x86_64/x86_64:/run/opengl-driver/lib/x86_64:/run/opengl-driver/lib/x86_64:/run/opengl-driver/lib:/run/opengl-driver-32/lib/tls/x86_64/x86_64:/run/opengl-driver-32/lib/tls/x86_64:/run/opengl-driver-32/lib/tls/x86_64:/run/opengl-driver-32/lib/tls:/run/opengl-driver-32/lib/x86_64/x86_64:/run/opengl-driver-32/lib/x86_64:/run/opengl-driver-32/lib/x86_64:/run/opengl-driver-32/lib(LD_LIBRARY_PATH)
      2630:       trying file=/nix/store/833lmkfczmbzmxfnvhb00bl0kmb9yly3-SDL-1.2.15/lib/libGL.so.1
      2630:       trying file=/nix/store/8fz98xhlk563kgsvmvihyj8m6i22pcc7-libogg-1.3.3/lib/tls/x86_64/x86_64/libGL.so.1
      2630:       trying file=/nix/store/8fz98xhlk563kgsvmvihyj8m6i22pcc7-libogg-1.3.3/lib/tls/x86_64/libGL.so.1
      2630:       trying file=/nix/store/8fz98xhlk563kgsvmvihyj8m6i22pcc7-libogg-1.3.3/lib/tls/x86_64/libGL.so.1
      2630:       trying file=/nix/store/8fz98xhlk563kgsvmvihyj8m6i22pcc7-libogg-1.3.3/lib/tls/libGL.so.1
      2630:       trying file=/nix/store/8fz98xhlk563kgsvmvihyj8m6i22pcc7-libogg-1.3.3/lib/x86_64/x86_64/libGL.so.1
      2630:       trying file=/nix/store/8fz98xhlk563kgsvmvihyj8m6i22pcc7-libogg-1.3.3/lib/x86_64/libGL.so.1
      2630:       trying file=/nix/store/8fz98xhlk563kgsvmvihyj8m6i22pcc7-libogg-1.3.3/lib/x86_64/libGL.so.1
      2630:       trying file=/nix/store/8fz98xhlk563kgsvmvihyj8m6i22pcc7-libogg-1.3.3/lib/libGL.so.1
      2630:       trying file=/nix/store/fw6kv6w6fjh5q7pmnr470pkp93ds13hb-libvorbis-1.3.6/lib/tls/x86_64/x86_64/libGL.so.1
      2630:       trying file=/nix/store/fw6kv6w6fjh5q7pmnr470pkp93ds13hb-libvorbis-1.3.6/lib/tls/x86_64/libGL.so.1
      2630:       trying file=/nix/store/fw6kv6w6fjh5q7pmnr470pkp93ds13hb-libvorbis-1.3.6/lib/tls/x86_64/libGL.so.1
      2630:       trying file=/nix/store/fw6kv6w6fjh5q7pmnr470pkp93ds13hb-libvorbis-1.3.6/lib/tls/libGL.so.1
      2630:       trying file=/nix/store/fw6kv6w6fjh5q7pmnr470pkp93ds13hb-libvorbis-1.3.6/lib/x86_64/x86_64/libGL.so.1
      2630:       trying file=/nix/store/fw6kv6w6fjh5q7pmnr470pkp93ds13hb-libvorbis-1.3.6/lib/x86_64/libGL.so.1
      2630:       trying file=/nix/store/fw6kv6w6fjh5q7pmnr470pkp93ds13hb-libvorbis-1.3.6/lib/x86_64/libGL.so.1
      2630:       trying file=/nix/store/fw6kv6w6fjh5q7pmnr470pkp93ds13hb-libvorbis-1.3.6/lib/libGL.so.1
      2630:       trying file=/nix/store/9c9vryq51hlw0jr8ddk2yk1qvkaq0vl2-curl-7.61.0/lib/tls/x86_64/x86_64/libGL.so.1
      2630:       trying file=/nix/store/9c9vryq51hlw0jr8ddk2yk1qvkaq0vl2-curl-7.61.0/lib/tls/x86_64/libGL.so.1
      2630:       trying file=/nix/store/9c9vryq51hlw0jr8ddk2yk1qvkaq0vl2-curl-7.61.0/lib/tls/x86_64/libGL.so.1
      2630:       trying file=/nix/store/9c9vryq51hlw0jr8ddk2yk1qvkaq0vl2-curl-7.61.0/lib/tls/libGL.so.1
      2630:       trying file=/nix/store/9c9vryq51hlw0jr8ddk2yk1qvkaq0vl2-curl-7.61.0/lib/x86_64/x86_64/libGL.so.1
      2630:       trying file=/nix/store/9c9vryq51hlw0jr8ddk2yk1qvkaq0vl2-curl-7.61.0/lib/x86_64/libGL.so.1
      2630:       trying file=/nix/store/9c9vryq51hlw0jr8ddk2yk1qvkaq0vl2-curl-7.61.0/lib/x86_64/libGL.so.1
      2630:       trying file=/nix/store/9c9vryq51hlw0jr8ddk2yk1qvkaq0vl2-curl-7.61.0/lib/libGL.so.1
      2630:       trying file=/nix/store/mb471p2iqq73lfz9z41p7gbszpa3rk9r-sane-config/lib/sane/tls/x86_64/x86_64/libGL.so.1
      2630:       trying file=/nix/store/mb471p2iqq73lfz9z41p7gbszpa3rk9r-sane-config/lib/sane/tls/x86_64/libGL.so.1
      2630:       trying file=/nix/store/mb471p2iqq73lfz9z41p7gbszpa3rk9r-sane-config/lib/sane/tls/x86_64/libGL.so.1
      2630:       trying file=/nix/store/mb471p2iqq73lfz9z41p7gbszpa3rk9r-sane-config/lib/sane/tls/libGL.so.1
      2630:       trying file=/nix/store/mb471p2iqq73lfz9z41p7gbszpa3rk9r-sane-config/lib/sane/x86_64/x86_64/libGL.so.1
      2630:       trying file=/nix/store/mb471p2iqq73lfz9z41p7gbszpa3rk9r-sane-config/lib/sane/x86_64/libGL.so.1
      2630:       trying file=/nix/store/mb471p2iqq73lfz9z41p7gbszpa3rk9r-sane-config/lib/sane/x86_64/libGL.so.1
      2630:       trying file=/nix/store/mb471p2iqq73lfz9z41p7gbszpa3rk9r-sane-config/lib/sane/libGL.so.1
      2630:       trying file=/run/opengl-driver/lib/tls/x86_64/x86_64/libGL.so.1
      2630:       trying file=/run/opengl-driver/lib/tls/x86_64/libGL.so.1
      2630:       trying file=/run/opengl-driver/lib/tls/x86_64/libGL.so.1
      2630:       trying file=/run/opengl-driver/lib/tls/libGL.so.1
      2630:       trying file=/run/opengl-driver/lib/x86_64/x86_64/libGL.so.1
      2630:       trying file=/run/opengl-driver/lib/x86_64/libGL.so.1
      2630:       trying file=/run/opengl-driver/lib/x86_64/libGL.so.1
      2630:       trying file=/run/opengl-driver/lib/libGL.so.1
      2630:       trying file=/run/opengl-driver-32/lib/tls/x86_64/x86_64/libGL.so.1
      2630:       trying file=/run/opengl-driver-32/lib/tls/x86_64/libGL.so.1
      2630:       trying file=/run/opengl-driver-32/lib/tls/x86_64/libGL.so.1
      2630:       trying file=/run/opengl-driver-32/lib/tls/libGL.so.1
      2630:       trying file=/run/opengl-driver-32/lib/x86_64/x86_64/libGL.so.1
      2630:       trying file=/run/opengl-driver-32/lib/x86_64/libGL.so.1
      2630:       trying file=/run/opengl-driver-32/lib/x86_64/libGL.so.1
      2630:       trying file=/run/opengl-driver-32/lib/libGL.so.1
      2630:      search cache=/nix/store/fg4yq8i8wd08xg3fy58l6q73cjy8hjr2-glibc-2.27/etc/ld.so.cache
      2630:      search path=/nix/store/fg4yq8i8wd08xg3fy58l6q73cjy8hjr2-glibc-2.27/lib/tls/x86_64/x86_64:/nix/store/fg4yq8i8wd08xg3fy58l6q73cjy8hjr2-glibc-2.27/lib/tls/x86_64:/nix/store/fg4yq8i8wd08xg3fy58l6q73cjy8hjr2-glibc-2.27/lib/tls/x86_64:/nix/store/fg4yq8i8wd08xg3fy58l6q73cjy8hjr2-glibc-2.27/lib/tls:/nix/store/fg4yq8i8wd08xg3fy58l6q73cjy8hjr2-glibc-2.27/lib/x86_64/x86_64:/nix/store/fg4yq8i8wd08xg3fy58l6q73cjy8hjr2-glibc-2.27/lib/x86_64:/nix/store/fg4yq8i8wd08xg3fy58l6q73cjy8hjr2-glibc-2.27/lib/x86_64:/nix/store/fg4yq8i8wd08xg3fy58l6q73cjy8hjr2-glibc-2.27/lib            (system search path)
      2630:       trying file=/nix/store/fg4yq8i8wd08xg3fy58l6q73cjy8hjr2-glibc-2.27/lib/tls/x86_64/x86_64/libGL.so.1
      2630:       trying file=/nix/store/fg4yq8i8wd08xg3fy58l6q73cjy8hjr2-glibc-2.27/lib/tls/x86_64/libGL.so.1
      2630:       trying file=/nix/store/fg4yq8i8wd08xg3fy58l6q73cjy8hjr2-glibc-2.27/lib/tls/x86_64/libGL.so.1
      2630:       trying file=/nix/store/fg4yq8i8wd08xg3fy58l6q73cjy8hjr2-glibc-2.27/lib/tls/libGL.so.1
      2630:       trying file=/nix/store/fg4yq8i8wd08xg3fy58l6q73cjy8hjr2-glibc-2.27/lib/x86_64/x86_64/libGL.so.1
      2630:       trying file=/nix/store/fg4yq8i8wd08xg3fy58l6q73cjy8hjr2-glibc-2.27/lib/x86_64/libGL.so.1
      2630:       trying file=/nix/store/fg4yq8i8wd08xg3fy58l6q73cjy8hjr2-glibc-2.27/lib/x86_64/libGL.so.1
      2630:       trying file=/nix/store/fg4yq8i8wd08xg3fy58l6q73cjy8hjr2-glibc-2.27/lib/libGL.so.1
      2630:
/nix/store/a1jmb86qk6fj3zvdmixhpwxlv1phliw6-openarena-0.8.8/openarena-0.8.8/openarena.x86_64: error while loading shared libraries: libGL.so.1: cannot open shared object file: No such file or directory
```

Same for minetest package:
```
$ minetest
libGL error: unable to load driver: radeonsi_dri.so
libGL error: driver pointer missing
libGL error: failed to load driver: radeonsi
libGL error: unable to load driver: radeonsi_dri.so
libGL error: driver pointer missing
libGL error: failed to load driver: radeonsi
libGL error: unable to load driver: swrast_dri.so
libGL error: failed to load driver: swrast
X Error of failed request:  BadValue (integer parameter out of range for operation)
  Major opcode of failed request:  156 (GLX)
  Minor opcode of failed request:  24 (X_GLXCreateNewContext)
  Value in failed request:  0x0
  Serial number of failed request:  45
  Current serial number in output stream:  46
```

I would appreciate if someone can explain how to point these to correct library or recompile, checking the store and pushing these off and back from configuration.nix was fruitless.

avnik commented:
But minetest works on master, so it definelly different problem

joepie91 commented:
This change seems to have broken most all OpenGL-based things in Rust, given that `glutin` [hardcodes the library names](https://github.com/tomaka/glutin/blob/c0b36056bdacdcef989f518d6d17307ac85f0da2/src/api/egl/mod.rs#L40) - this is rather inconvenient when developing things, since this seems to make it impossible to use `nix-shell` alongside Cargo to write and test Rust graphics code.

jb55 commented:
> If there's a static executable that wants to work on all Linux platforms, how can it find OpenGL in such a way that it will find the correct version on NixOS?

I think this is a valid point. I'm working on a game and I'm looking to do static linking to maximize portability between distros. A common practice when distributing games is to dlopen libGL to fetch all the available opengl function pointers.

It seems like this assumption is no longer valid? If I was a game developer attempting to target most linux distros what would I do? Please don't say steam runtime :|

jb55 commented:
I think I was confused, it doesn't look like you can do what I was looking to do. I think if I just link SDL2 and use the load functions from there it should be fine...

avnik commented:
@jb55 at least we have both SDL and SDL2 correctly patched to find all libraries in nixos in proper places. If you plan closed source game -- maybe env variable  can be good level of indirection?
